src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+   public static getId(): string {
+      return 'pep-avatars';
+   }
+
+   public static getName(): string {
+      return 'PEP-based Avatars';
+   }
+
+   public static getMetaData(): IMetaData {
+      return {
+         description: Translation.t('setting-pep-avatar-enable'),
+         xeps: [
+            {
+               id: 'XEP-0084',
+               name: 'User Avatar',
+               version: '1.1.4',
+            },
+         ],
+      };
+   }
+
+   constructor(pluginAPI: PluginAPI) {
+      super(MIN_VERSION, MAX_VERSION, pluginAPI);
+
+      let connection = pluginAPI.getConnection();
+
+      connection.registerHandler(this.onMessageAvatarUpdate, 'http://jabber.org/protocol/pubsub#event', 'message');
+
+      pluginAPI.addAvatarProcessor(this.avatarProcessor, 49);
+      pluginAPI.addPublishAvatarProcessor(this.publishAvatarProcessor, 49);
+      pluginAPI.addFeature(UANS_NOTIFY);
+   }
+
+   public getStorage() {
+      return this.pluginAPI.getStorage();
+   }
+
+   private publishAvatarProcessor = (avatar: IAvatar | null): Promise<[IAvatar]> => {
+      let connection = this.pluginAPI.getConnection();
+
+      if (!avatar || avatar.getData() === undefined) {
+         let item = $build('metadata', { xmlns: UANS_METADATA });
+
+         return connection
+            .getPEPService()
+            .publish(UANS_METADATA, item.tree(), UANS_METADATA)
+            .then(function (result) {
+               if ($(result).attr('type') === 'result') {
+                  return [undefined];
+               } else {
+                  return [avatar];
+               }
+            });
+      } else {
+         let data = $build('data', { xmlns: UANS_DATA }).t(avatar.getData()).tree();
+
+         return connection
+            .getPEPService()
+            .publish(UANS_DATA, data, UANS_DATA)
+            .then((result: any) => {
+               if ($(result).attr('type') === 'result') {
+                  let metadata;
+                  let hash = Hash.SHA1FromBase64(avatar.getData());
+                  let i = new Image();
+                  let iheight = 0;
+                  let iwidth = 0;
+                  let size = FileHelper.getFileSizeFromBase64(avatar.getData());
+
+                  i.onload = function () {
+                     iheight = i.height;
+                     iwidth = i.width;
+                  };
+                  i.src = 'data:' + avatar.getType() + ';base64,' + avatar.getData();
+                  metadata = $build('metadata', { xmlns: UANS_METADATA })
+                     .c('info', { bytes: size, id: hash, height: iheight, width: iwidth, type: avatar.getType() })
+                     .tree();
+
+                  return connection
+                     .getPEPService()
+                     .publish(UANS_METADATA, metadata, UANS_METADATA)
+                     .then(function (result) {
+                        if ($(result).attr('type') === 'result') {
+                           return [undefined];
```

Comment:
This would prevent other plugins to update the avatar as well.

Reply:
why?

Reply:
Because in the pipe every processor is called one after the other and the return value is always the "input" value of the next processor. Therefore the next processor will think that the avatar was deleted.

Reply:
so we should return avatar variable then..

Reply:
changed...